### ISSUE TITLE
[spirv] RelaxedPrecision decorations are missing in almost all cases when using min precision types

### ISSUE THREAD

**Issue by chrvoss** [Microsoft/DirectXShaderCompiler#1736]
Note this is **without** the -enable-16bit-types option.

As per https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/SPIR-V.rst#minimal-precision-scalar-types, using `min16float` in HLSL should be emitted as `OpTypeFloat 32` with the `RelaxedPrecision` decoration. However this decoration seems to be missing in almost all cases.

Specifically:

- Shader inputs and outputs are not decorated
- Results of expressions operation on `RelaxedPrecision` values are not decorated as `RelaxedPrecision` themselves

In fact, it appears the only way to generate intermediates with `RelaxedPrecision` at all is as the result of a phi expression.

Repro example:

```
// Compile with .\dxc -spirv -T ps_6_0 test.hlsl
// Tested against 3a3d16435fb493fc770c816dad6a576ef792c36c
min16float main(min16float x : TEXCOORD0) : SV_Target
{
    min16float y;
    if(x > 0)
    {
        y = x * x;
    }
    else
    {
        y = x;
    }
    return y * y;
}
```

This results in the following output:

```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 20
; Schema: 0
               OpCapability Shader
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %main "main" %in_var_TEXCOORD0 %out_var_SV_Target
               OpExecutionMode %main OriginUpperLeft
               OpSource HLSL 620
               OpName %main "main"
               OpName %in_var_TEXCOORD0 "in.var.TEXCOORD0"
               OpName %out_var_SV_Target "out.var.SV_Target"
               OpDecorate %in_var_TEXCOORD0 Location 0
               OpDecorate %out_var_SV_Target Location 0
               OpDecorate %4 RelaxedPrecision
       %void = OpTypeVoid
          %6 = OpTypeFunction %void
      %float = OpTypeFloat 32
%_ptr_Input_float = OpTypePointer Input %float
%_ptr_Output_float = OpTypePointer Output %float
    %float_0 = OpConstant %float 0
       %bool = OpTypeBool
%in_var_TEXCOORD0 = OpVariable %_ptr_Input_float Input
%out_var_SV_Target = OpVariable %_ptr_Output_float Output
       %main = OpFunction %void None %6
         %12 = OpLabel
         %13 = OpLoad %float %in_var_TEXCOORD0
         %14 = OpFOrdGreaterThan %bool %13 %float_0
               OpSelectionMerge %15 None
               OpBranchConditional %14 %16 %17
         %16 = OpLabel
         %18 = OpFMul %float %13 %13
               OpBranch %15
         %17 = OpLabel
               OpBranch %15
         %15 = OpLabel
          %4 = OpPhi %float %18 %16 %13 %17
         %19 = OpFMul %float %4 %4
               OpStore %out_var_SV_Target %19
               OpReturn
```

In this example `%4` is the only decorated value, but the next operation assigning to `%19` immediately loses the decoration again. The expected result would be that `%in_var_TEXCOORD0`, `%out_var_SV_Target`, as well as intermediates `%13`, `%18` and `%19` should have the decoration as well.

While investigating, I noticed that `SPIRVEmitter::processBinaryOp` has this code:

```
    // Propagate RelaxedPrecision
    if (lhsVal.isRelaxedPrecision() || rhsVal.isRelaxedPrecision())
      result.setRelaxedPrecision();
```

However this code doesn't tie into the rest of the codegen, as `SpirvEvalInfo::setRelaxedPrecision` and `SpirvEvalInfo::isRelaxedPrecision` aren't called anywhere else.

**Comment by ehsannas** [comment]
In short: This is a known issue and we are working on it! :-)

Hi  @chrvoss , Thanks for reporting. RelaxedPrecision is currently not fully supported as you pointed out: the current code does not propagate this decoration forward.

This is due to the shortcomings of the current SPIR-V CodeGen infrastructure. We are working on improving the infrastructure code that's needed to support such propagations. It might take a bit of time for us to finalize the code and land it.

**Comment by mmostajab** [comment]
Hi @ehsannas 

Thanks for your efforts. 

Is there any update on this ticket? Is it supported properly now? I remember I tried it in late November but I could barely get RelaxedPrecision decoration and that is very important when using shaders on mobile chipsets.

**Comment by Cry-Mory** [comment]
Thanks a lot, @ehsannas for your quick reply.

Looking forward to the new features.

**Comment by ehsannas** [comment]
I have a working prototype [here](https://github.com/ehsannas/DirectXShaderCompiler/tree/propagate_relaxed_precision).
Landing the change is currently blocked by PR #1851 .

**Comment by Cry-Mory** [comment]
@ehsannas Thanks a lot. I will give this a try, soon and let you know about the results we get here.

**Comment by Cry-Mory** [comment]
Hi @ehsannas,

Thanks for providing us a prototype of relaxed precision support. We tested it and it is clearly producing more relaxed precision decorated operations than before. We also profiled it a bit and found out relaxed precision decorated operations are improving performance.

Meanwhile, we found two issues which I would like to mention. So, they may help you to make some improvements:

**1. Operations involving two operands which one is a float variable, and the other is min16float**
An operation which involves two operands that one of them is of type _float_ and the other one is decorated with RelaxedPrecision, produces RelaxedPrecision decorated result. As an example, we were subtracting a float variable from a min16float typed variable, and the result was decorated with relaxed precision float. This may cause issue on architectures using 16-bit floats; Particularly if result does not fit in16-bit.

See DXC output below:

> **For:**
>    float           a = float(IN.WPos.x);
>    min16float b = float(IN.WPos.y);
>    OUT.Color.rgba = a - b;
> **Produces:**
> OpDecorate %5 RelaxedPrecision
> %5 = OpFSub %float %14 %15


**2. SampleCmpLevelZero does not get relaxed precision decorator**
According to [Spriv's relaxed precision documentation](https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#_a_id_relaxedprecisionsection_a_relaxed_precision) and [Spirv's OpImageSampleDrefExplicitLod documentation](https://www.khronos.org/registry/spir-v/specs/1.0/SPIRV.html#OpImageSampleDrefExplicitLod) documentation, a OpImageSampleDrefExplicitLod can be decorated with relaxed precision (OpImageSampleDrefExplicitLod maps to _SampleCmpLevelZero_ in hlsl) but DXC is not doing it. See below:

> **For:**
  >OUT.Color.rgba = colorChartTex.SampleCmpLevelZero(linearCompareSS, float2(0,0), 0);
  >**Produces:**
  >%25 = OpImageSampleDrefExplicitLod %float %24 %15 %float_0 Lod %float_0

We expected that %25 to have RelaxedPrecision decorator. We tried the same situation with hlsl's texture's _Sample_ function call and it produces relaxed precision decorator for sampling (as we expected):
> **For:**
  >OUT.Color.rgba = colorChartTex.Sample(linearClampSS, float2(0,0));
  >**Produces:**
  >OpDecorate %11 RelaxedPrecision
  >%11 = OpImageSampleImplicitLod %v4float %26 %17 None

You can reproduce the mentioned issues with the following shader:
[TestShader.txt](https://github.com/Microsoft/DirectXShaderCompiler/files/2895515/TestShader.txt)

By default SampleCmpLevelZero is reproduced with this shader but if you open it and comment and uncomment specific lines (they are clearly visible), you can reproduce the other issue.

To compile it with DXC you can use the following command line:

> dxc.exe -fvk-use-dx-position-w  -O3 -Zpr -spirv -no-warnings -E TonemappingMobilePS -T ps_5_0 TestShader.txt -Fh TestShader.h_spv

Thanks in advanced.

**Comment by ehsannas** [comment]
Hi @Cry-Mory 
Thanks a lot for your detailed report. I'm glad to hear the PR makes a positive difference so far. I'll look into the issues you mentioned and update the PR before landing. Thanks for testing it out.

**Comment by ehsannas** [comment]
Regarding issue #1: You're right. My current implementation propagates the decoration even if only 1 of the arguments is a min type. I get your point, and I think it should be fixed.

What makes it a bit more complicated is cases where a literal is provided.
```hlsl
min16float b = float(IN.WPos.y);
OUT.Color.rgba = b - 1.5;
```
the literal in the source code has an unknown bitwidth (it could a 32-bit `1.5` or a 16-bit `1.5` or a 64-bit `1.5`). So the judgement of whether or not we should add the decoration to the OpFSub in this case is not quite clear. Although one possible solution for this case would be to try to figure out if the given literal number can be fit in 16 bits, and if so, use relaxed precision.
For example: `1.234534564573423426` may not fit in 16-bits (😄) and therefore if it's used in the subtraction above, we should not use relaxed precision.

**Comment by Cry-Mory** [comment]
@ehsannas I think that is a good idea. Maybe adding a suffix to these literals to force them to be 16-bit is not a bad idea (at least I am not aware of having any).

**Comment by ehsannas** [comment]
@Cry-Mory I have posted a [fix](https://github.com/Microsoft/DirectXShaderCompiler/pull/1915/commits/ec83cd4eaa52d9a4b60e9e88002f95ea725f2748) to PR #1915 which should address both of the improvements you mentioned above. Please feel free to test it out :)

**Comment by ehsannas** [comment]
Landed. Feel free to reopen this issue, or file a new issue if you have any other concerns. Thanks.
